src/internal/pjsdb/job_iterator.go

```diff
@@ -0,0 +1,131 @@
+// keeping this as its own separate struct makes it easy to check to see if a
+// filter is defined. It also allows filters to be combined later if needed.
+type IterateJobsFilter struct {
+	Parent   JobID
+	Input    []byte
+	Spec     []byte
+	SpecHash []byte
+	Output   []byte
+	Error    string
+
+	// TODO(Fahad): think through how to support querying on times.
+	//Queued     time.Time
+	//Processing time.Time
+	//Done       time.Time
+}
+
+func (f *IterateJobsFilter) apply() (conditions []string, values []any) {
+	if f.Parent != 0 {
+		conditions = append(conditions, "parent = ?")
+		values = append(values, f.Parent)
+	}
+	if f.Spec != nil {
+		conditions = append(conditions, "spec = ?")
+		values = append(values, f.Spec)
+	}
+	if f.Input != nil {
+		conditions = append(conditions, "input = ?")
+		values = append(values, f.Input)
+	}
+	if f.SpecHash != nil {
+		conditions = append(conditions, "spec_hash = ?")
+		values = append(values, f.SpecHash)
+	}
+	if f.Output != nil {
+		conditions = append(conditions, "output = ?")
+		values = append(values, f.Output)
+	}
+	if f.Error != "" {
+		conditions = append(conditions, "error = ?")
+		values = append(values, f.Error)
+	}
+	return conditions, values
+}
+
+type IterateJobsRequest struct {
+	IteratorConfiguration
+	Filter *IterateJobsFilter
+}
+
+// JobIterator implements stream.Iterator[T] on Job objects following the
+// pattern set by the pfsdb package.
+// Iterators are the underlying type used by list-style crud operations.
+type JobIterator struct {
+	paginator pageIterator[JobRow]
+	extCtx    sqlx.ExtContext
+}
+
+var _ stream.Iterator[Job] = &JobIterator{} // catch changes that break the interface.
+
+func (i *JobIterator) Next(ctx context.Context, dst *Job) error {
+	if dst == nil {
+		return errors.Errorf("dst JobRow cannot be nil")
+	}
+	row, err := i.paginator.next(ctx, i.extCtx)
+	if err != nil {
+		return err
+	}
+	job := &Job{JobRow: *row}
+	*dst = *job
+	return nil
+}
+
+func NewJobsIterator(extCtx sqlx.ExtContext, req IterateJobsRequest) *JobIterator {
+	var conditions []string
+	var values []any
+	query := `SELECT * FROM pjs.jobs`
+	if req.Filter != nil {
+		conditions, values = req.Filter.apply()
+	}
+	if len(conditions) > 0 {
+		query += "\n" + fmt.Sprintf("WHERE %s", strings.Join(conditions, " AND "))
+	}
+	query += req.orderBy()
+	query = extCtx.Rebind(query)
+	if req.PageSize == 0 {
+		req.PageSize = defaultPageSize
+	}
+	return &JobIterator{
+		paginator: newPageIterator[JobRow](query, values, req.StartPage, req.PageSize, 0),
+		extCtx:    extCtx,
+	}
+}
+
+// ForEachJob iterates over each Job 'job' in the jobs table and executes the callback cb(job).
+func ForEachJob(ctx context.Context, db *pachsql.DB, req IterateJobsRequest, cb func(job Job) error) error {
+	ctx = pctx.Child(ctx, "forEachJob")
+	if err := stream.ForEach[Job](pctx.Child(ctx, "forEach"), NewJobsIterator(db, req), cb); err != nil {
```

Comment:
Is there a need to make a child context immediately after making a child context?